src/docfx/cli/CommandLine.cs

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.CommandLine;
+using System.CommandLine.Invocation;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Docs.Build
+{
+    internal class CommandLine
+    {
+        private string _command = string.Empty;
+
+        private CommandLineOptions _options = new CommandLineOptions();
+
+        public static (string command, CommandLineOptions options) Parse(string[] args)
+            => new CommandLine(args).Command;
+
+        private CommandLine(string[] args)
+        {
+            if (args.Length == 0)
+            {
+                // Show usage when just running `docfx`
+                args = new[] { "--help" };
+            }
+
+            try
+            {
+                new RootCommand
+                {
+                    NewCommand(),
+                    RestoreCommand(),
+                    BuildCommand(),
+                    ServeCommand(),
+                }
+                .Invoke(args);
+                if (_options.Stdin && Console.ReadLine() is string stdin)
+                {
+                    _options.StdinConfig = JsonUtility.DeserializeData<JObject>(stdin, new FilePath("--stdin"));
+                }
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex.Message);
+            }
+        }
+
+        private (string command, CommandLineOptions options) Command
+            => (_command, _options);
+
+        private Command BaseCommand(string name)
+            => new Command(name)
+            {
+                Handler = CommandHandler.Create((CommandLineOptions options) => Parse(name, options)),
```

Comment:
consider move this to `docfx.cs` and hook the handler to the corresponding function directly.